File: composeApp/src/androidMain/kotlin/com/dnfapps/arrmatey/ui/viewmodel/SonarrViewModel.kt
package com.dnfapps.arrmatey.ui.viewmodel

import androidx.lifecycle.viewModelScope
import com.dnfapps.arrmatey.api.arr.model.ArrSeries
import com.dnfapps.arrmatey.api.arr.model.CommandPayload
import com.dnfapps.arrmatey.api.arr.viewmodel.SonarrRepository
import com.dnfapps.arrmatey.model.Instance
import kotlinx.coroutines.launch
import kotlin.time.ExperimentalTime

class SonarrViewModel(instance: Instance): ArrViewModel(instance) {

    private val sonarrRepo = repository as SonarrRepository

    val episodeState = sonarrRepo.episodeState

    fun getEpisodes(seriesId: Int, seasonNumber: Int? = null) {
        viewModelScope.launch {
            sonarrRepo.getEpisodes(seriesId, seasonNumber)
        }
    }

    override fun getDetails(id: Int) {
        super.getDetails(id)
        getEpisodes(id)
    }

    override fun searchPayload(ids: List<Int>): CommandPayload {
        return CommandPayload.SonarrSearch(ids)
    }

    @OptIn(ExperimentalTime::class)
    fun toggleSeasonMonitor(series: ArrSeries, seasonNumber: Int) {
        viewModelScope.launch {
            sonarrRepo.toggleSeasonMonitorState(series, seasonNumber)
        }
    }

    fun toggleEpisodeMonitor(episodeId: Long) {
        viewModelScope.launch {
            sonarrRepo.toggleEpisodeMonitorState(episodeId)
        }
    }

}

File: iosApp/iosApp/ViewModel/SonarrViewModel.swift
//
//  SonarrViewModel.swift
//  iosApp
//
//  Created by Owen LeJeune on 2025-12-15.
//

import Shared

class SonarrViewModel: ArrViewModel {
    var repository: any IArrRepository
    
    let instance: Instance

    private var sonarrRepository: SonarrRepository {
        repository as! SonarrRepository
    }
    
    init(instance: Instance) {
        self.instance = instance
        self.repository = SonarrRepository(instance: instance)
    }
    
    func getEpisodeState() -> SkieSwiftStateFlow<EpisodeUiState> {
        return sonarrRepository.episodeState
    }
    
    func getEpsiodes(seriesId: Int32, seasonNumber: Int32? = nil) async {
        do {
            try await sonarrRepository.getEpisodes(seriesId: seriesId, seasonNumber: seasonNumber as? KotlinInt)
        } catch {
            return
        }
    }
    
    func toggleSeasonMonitor(series: ArrSeries, seasonNumber: Int32) async {
        do {
            try await sonarrRepository.toggleSeasonMonitorState(series: series, seasonNumber: seasonNumber)
        } catch {
            return
        }
    }
    
    func toggleEpisodeMonitor(episodeId: Int64) async {
        do {
            try await sonarrRepository.toggleEpisodeMonitorState(episodeId: episodeId)
        } catch {
            return
        }
    }
}
